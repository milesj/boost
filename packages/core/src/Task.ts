/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Context from './Context';
import {
  STATUS_PENDING,
  STATUS_RUNNING,
  STATUS_SKIPPED,
  STATUS_PASSED,
  STATUS_FAILED,
} from './constants';
import { Status } from './types';
import Emitter from './Emitter';

export type TaskAction<Ctx extends Context> = (
  context: Ctx,
  value: any,
  task: Task<Ctx>,
) => any | Promise<any>;

export interface TaskMetadata {
  depth: number;
  index: number;
  startTime: number;
  stopTime: number;
}

export default class Task<Ctx extends Context> extends Emitter {
  action: TaskAction<Ctx>;

  // @ts-ignore Set after instantiation
  context: Ctx;

  title: string;

  metadata: TaskMetadata = {
    depth: 0,
    index: 0,
    startTime: 0,
    stopTime: 0,
  };

  output: string = '';

  parent: Task<Ctx> | null = null;

  status: Status = STATUS_PENDING;

  statusText: string = '';

  tasks: Task<Ctx>[] = [];

  constructor(title: string, action: TaskAction<Ctx>) {
    super();

    if (!title || typeof title !== 'string') {
      throw new Error('Tasks require a title.');
    }

    if (action !== null && typeof action !== 'function') {
      throw new Error('Tasks require an executable function.');
    }

    this.action = action;
    this.status = STATUS_PENDING;
    this.title = title;
  }

  /**
   * Return true if the task failed when executing.
   */
  hasFailed(): boolean {
    return this.status === STATUS_FAILED;
  }

  /**
   * Return true if the task executed successfully.
   */
  hasPassed(): boolean {
    return this.status === STATUS_PASSED;
  }

  /**
   * Return true if the task has been completed in any form.
   */
  isComplete(): boolean {
    return this.hasPassed() || this.hasFailed() || this.isSkipped();
  }

  /**
   * Return true if the task has not been executed yet.
   */
  isPending(): boolean {
    return this.status === STATUS_PENDING;
  }

  /**
   * Return true if the task is currently running.
   */
  isRunning(): boolean {
    return this.status === STATUS_RUNNING;
  }

  /**
   * Return true if the task was or will be skipped.
   */
  isSkipped(): boolean {
    return this.status === STATUS_SKIPPED;
  }

  /**
   * Run the current task by executing it and performing any before and after processes.
   */
  async run(context: Ctx, value: any): Promise<any> {
    this.setContext(context);
    this.emit('run', [value]);

    if (this.isSkipped()) {
      this.status = STATUS_SKIPPED;
      this.emit('skip', [value]);

      return Promise.resolve(value);
    }

    this.status = STATUS_RUNNING;
    this.metadata.startTime = Date.now();

    try {
      this.output = await this.action(context, value, this);
      this.status = STATUS_PASSED;
      this.metadata.stopTime = Date.now();
      this.emit('pass', [this.output]);
    } catch (error) {
      this.status = STATUS_FAILED;
      this.metadata.stopTime = Date.now();
      this.emit('fail', [error]);

      throw error;
    }

    this.statusText = '';

    return this.output;
  }

  /**
   * Set the context to be passed around.
   */
  setContext(context: Ctx): this {
    this.context = context;

    return this;
  }

  /**
   * Mark a task as skipped if the condition is true.
   */
  skip(condition: boolean = true): this {
    if (condition) {
      this.status = STATUS_SKIPPED;
    }

    return this;
  }
}
