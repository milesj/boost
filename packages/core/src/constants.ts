import { Status } from './types';

export const APP_NAME_PATTERN = /^[a-z]{1}[-a-z0-9]+[a-z]{1}$/u;
export const CONFIG_NAME_PATTERN = /^[a-z]{1}[a-zA-Z0-9]+$/u;
export const EVENT_NAME_PATTERN = /^[-a-z.]+$/u;
export const MODULE_NAME_PATTERN = /^(@[-a-z]+\/)?[-a-z]+$/u;
export const PLUGIN_NAME_PATTERN = /^([a-z]+):[-a-z]+$/u;

export const STATUS_PENDING: Status = 'pending';
export const STATUS_RUNNING: Status = 'running';
export const STATUS_SKIPPED: Status = 'skipped';
export const STATUS_PASSED: Status = 'passed';
export const STATUS_FAILED: Status = 'failed';

// Chalk colors
// yellow = app name, module name
// cyan = file
// magenta = plugin name
