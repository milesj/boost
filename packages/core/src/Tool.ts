/**
 * @copyright   2017-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

/* eslint-disable no-param-reassign */

import path from 'path';
import util from 'util';
import chalk from 'chalk';
import debug from 'debug';
import pluralize from 'pluralize';
import i18next from 'i18next';
import mergeWith from 'lodash/mergeWith';
import optimal, { bool, object, string, Blueprint } from 'optimal';
import parseArgs, { Arguments, Options as ArgOptions } from 'yargs-parser';
import ConfigLoader from './ConfigLoader';
import Console from './Console';
import Emitter from './Emitter';
import ModuleLoader, { Constructor } from './ModuleLoader';
import Plugin from './Plugin';
import Reporter from './Reporter';
import DefaultReporter from './reporters/DefaultReporter';
import ErrorReporter from './reporters/ErrorReporter';
import enableDebug from './helpers/enableDebug';
import handleMerge from './helpers/handleMerge';
import isEmptyObject from './helpers/isEmptyObject';
import CIReporter from './reporters/CIReporter';
import LanguageDetector from './i18n/LanguageDetector';
import FileBackend from './i18n/FileBackend';
import themePalettes from './themes';
import { DEFAULT_TOOL_CONFIG } from './constants';
import { Debugger, Translator, ToolConfig, PackageConfig } from './types';

export interface ToolOptions {
  appName: string;
  appPath: string;
  argOptions: ArgOptions;
  configBlueprint: Blueprint;
  configFolder: string;
  footer: string;
  header: string;
  root: string;
  scoped: boolean;
  workspaceRoot: string;
}

export interface PluginContract<T = any> {
  contract: Constructor<T>;
  loader: ModuleLoader<T>;
  pluralName: string;
  singularName: string;
}

export type PluginRegistry = { [type: string]: Plugin<any> };

export default class Tool<
  Config extends ToolConfig = ToolConfig,
  Registry extends PluginRegistry = PluginRegistry
> extends Emitter {
  args?: Arguments;

  argv: string[] = [];

  // @ts-ignore Allow default spread
  config: Config = { ...DEFAULT_TOOL_CONFIG };

  console: Console;

  debug: Debugger;

  options: ToolOptions;

  package: PackageConfig = { name: '' };

  plugins: { [K in keyof Registry]?: Registry[K][] } = {};

  pluginTypes: { [K in keyof Registry]?: PluginContract<Registry[K]> } = {};

  reporters: Reporter<any>[] = [];

  translator: Translator;

  private configLoader: ConfigLoader;

  private initialized: boolean = false;

  private reporterLoader: ModuleLoader<Reporter<any>>;

  constructor(options: Partial<ToolOptions>, argv: string[] = []) {
    super();

    this.argv = argv;

    this.options = optimal(
      options,
      {
        appName: string().required(),
        appPath: string().required(),
        argOptions: object(),
        configBlueprint: object(),
        configFolder: string('./configs'),
        footer: string().empty(),
        header: string().empty(),
        root: string(process.cwd()),
        scoped: bool(),
        workspaceRoot: string().empty(),
      },
      {
        name: 'Tool',
      },
    );

    // Core debugger for the entire tool
    this.debug = this.createDebugger('core');

    // Setup i18n translation
    this.translator = this.createTranslator();

    // Initialize the console first so we can start logging
    this.console = new Console(this);

    // Add a reporter to catch errors during initialization
    this.addReporter(new ErrorReporter());

    // Make these available for testing purposes
    this.configLoader = new ConfigLoader(this);
    this.reporterLoader = new ModuleLoader(this, 'reporter', Reporter, true);

    // Cleanup when an exit occurs
    /* istanbul ignore next */
    if (process.env.NODE_ENV !== 'test') {
      process.on('exit', code => {
        this.emit('exit', [code]);
      });
    }

    // eslint-disable-next-line global-require
    this.debug('Using boost v%s', require('../package.json').version);
  }

  /**
   * Add a plugin for a specific contract type and bootstrap with the tool.
   */
  addPlugin<T extends Plugin<any>>(typeName: keyof Registry, plugin: T): this {
    const type = this.pluginTypes[typeName];

    if (!type) {
      throw new Error(this.msg('errors:pluginContractNotFound', { typeName }));
    } else if (!(plugin instanceof Plugin)) {
      throw new TypeError(this.msg('errors:pluginNotExtended', { parent: 'Plugin', typeName }));
    } else if (!(plugin instanceof type.contract)) {
      throw new TypeError(
        this.msg('errors:pluginNotExtended', { parent: type.contract.name, typeName }),
      );
    }

    plugin.tool = this;
    plugin.bootstrap();

    if (this.plugins[typeName]) {
      this.plugins[typeName]!.push(plugin);
    } else {
      this.plugins[typeName] = [plugin];
    }

    return this;
  }

  /**
   * Add a reporter and bootstrap with the tool and console.
   */
  addReporter(reporter: Reporter<any>): this {
    reporter.console = this.console;
    reporter.tool = this;
    reporter.bootstrap();

    this.reporters.push(reporter);

    return this;
  }

  /**
   * Create a debugger instance with a namespace.
   */
  createDebugger(...namespaces: string[]): Debugger {
    const handler = debug(`${this.options.appName}:${namespaces.join(':')}`) as Debugger;

    handler.invariant = (condition: boolean, message: string, pass: string, fail) => {
      handler('%s: %s', message, condition ? chalk.green(pass) : chalk.red(fail));
    };

    return handler;
  }

  /**
   * Create an i18n translator instance.
   */
  createTranslator(): Translator {
    return i18next
      .createInstance()
      .use(new LanguageDetector())
      .use(new FileBackend())
      .init(
        {
          backend: {
            resourcePaths: [
              path.join(__dirname, 'resources'),
              path.join(this.options.appPath, 'resources'),
            ],
          },
          defaultNS: 'app',
          fallbackLng: ['en'],
          fallbackNS: 'common',
          initImmediate: false,
          lowerCaseLng: true,
          ns: ['app', 'common', 'errors', 'prompts'],
        },
        error => {
          if (error) {
            throw error;
          }
        },
      );
  }

  /**
   * Force exit the application.
   */
  exit(message: string | Error | null, code: number = 1): this {
    this.console.exit(message, code);

    return this;
  }

  /**
   * Get a plugin by name and type.
   */
  getPlugin<T extends Plugin<any>>(typeName: keyof Registry, name: string): T {
    if (!this.pluginTypes[typeName]) {
      throw new Error(this.msg('errors:pluginContractNotFound', { typeName }));
    }

    const plugin = this.plugins[typeName]!.find(p => p.name === name);

    if (plugin) {
      return plugin as T;
    }

    throw new Error(
      this.msg('errors:pluginNotFound', {
        name,
        typeName,
      }),
    );
  }

  /**
   * Get a reporter by name.
   */
  getReporter(name: string): Reporter<any> {
    const reporter = this.reporters.find(r => r.name === name);

    if (reporter) {
      return reporter;
    }

    throw new Error(this.msg('errors:reporterNotFound', { name }));
  }

  /**
   * Return a list of Boost bundled theme names.
   */
  getThemeList(): string[] {
    return Object.keys(themePalettes);
  }

  /**
   * Initialize the tool by loading config and plugins.
   */
  initialize(): this {
    if (this.initialized) {
      return this;
    }

    const { appName } = this.options;
    const pluginNames = Object.values(this.pluginTypes).map(type => type!.singularName);

    this.debug('Initializing %s', chalk.yellow(appName));

    this.args = parseArgs(
      this.argv,
      mergeWith(
        {
          array: ['extends', 'reporter', ...pluginNames],
          boolean: ['debug', 'silent'],
          number: ['output'],
          string: ['config', 'extends', 'locale', 'reporter', 'theme', ...pluginNames],
        },
        this.config.argOptions,
        handleMerge,
      ),
    );

    this.loadConfig();
    this.loadPlugins();
    this.loadReporters();

    this.initialized = true;

    return this;
  }

  /**
   * Load the package.json and local configuration files.
   *
   * Must be called first in the lifecycle.
   */
  loadConfig(): this {
    if (this.initialized) {
      return this;
    }

    this.package = this.configLoader.loadPackageJSON();
    this.config = this.configLoader.loadConfig(this.args!);

    // Inherit workspace metadata
    this.options.workspaceRoot = this.configLoader.workspaceRoot;

    // Enable debugger
    if (this.config.debug) {
      enableDebug(this.options.appName);
    }

    // Update locale
    if (this.config.locale) {
      this.translator.changeLanguage(this.config.locale);
    }

    return this;
  }

  /**
   * Register plugins from the loaded configuration.
   *
   * Must be called after config has been loaded.
   */
  loadPlugins(): this {
    if (this.initialized) {
      return this;
    }

    if (isEmptyObject(this.config)) {
      throw new Error(this.msg('errors:configNotLoaded'));
    }

    Object.keys(this.pluginTypes).forEach(type => {
      const { loader, pluralName } = this.pluginTypes[type as keyof Registry]!;
      const plugins = loader.loadModules(this.config[pluralName]);

      // Sort plugins by priority
      loader.debug('Sorting plugins by priority');

      plugins.sort((a, b) => a.priority - b.priority);

      // Bootstrap each plugin with the tool
      loader.debug('Bootstrapping plugins with tool environment');

      plugins.forEach(plugin => {
        this.addPlugin(type, plugin);
      });
    });

    return this;
  }

  /**
   * Register reporters from the loaded configuration.
   *
   * Must be called after config has been loaded.
   */
  loadReporters(): this {
    if (this.initialized) {
      return this;
    }

    if (isEmptyObject(this.config)) {
      throw new Error(this.msg('errors:configNotLoaded', { name: 'reporters' }));
    }

    const reporters: Reporter<any>[] = [];

    if (process.env.CI && !process.env.BOOST_ENV) {
      this.reporterLoader.debug(
        'CI environment detected, using %s CI reporter',
        chalk.yellow('boost'),
      );

      reporters.push(new CIReporter());
    } else {
      reporters.push(...this.reporterLoader.loadModules(this.config.reporters));
    }

    // Use default reporter
    if (reporters.length === 0) {
      this.reporterLoader.debug('Using default %s reporter', chalk.yellow('boost'));

      reporters.push(new DefaultReporter());
    }

    // Bootstrap each plugin with the tool
    this.reporterLoader.debug('Bootstrapping reporters with tool and console environment');

    reporters.forEach(reporter => {
      this.addReporter(reporter);
    });

    return this;
  }

  /**
   * Log a message to the console to display on success.
   */
  log(message: string, ...args: any[]): this {
    this.console.log(util.format(message, ...args));

    return this;
  }

  /**
   * Log an error to the console to display on failure.
   */
  logError(message: string, ...args: any[]): this {
    this.console.logError(util.format(message, ...args));

    return this;
  }

  /**
   * Retrieve a translated message from a resource bundle.
   */
  msg(key: string | string, params?: any, options?: i18next.TranslationOptions): string {
    return this.translator.t(key, {
      interpolation: { escapeValue: false },
      replace: params,
      ...options,
    });
  }

  /**
   * Register a custom type of plugin, with a defined contract all instances should extend.
   */
  registerPlugin(
    typeName: keyof Registry,
    contract: Constructor<any>,
    useBoost: boolean = false,
  ): this {
    if (this.pluginTypes[typeName]) {
      throw new Error(this.msg('errors:pluginContractExists', { typeName }));
    }

    const name = String(typeName);

    this.debug('Registering new plugin type: %s', chalk.green(name));

    this.plugins[typeName] = [];

    this.pluginTypes[typeName] = {
      contract,
      loader: new ModuleLoader(this, name, contract, useBoost),
      pluralName: pluralize(name),
      singularName: name,
    };

    return this;
  }
}
