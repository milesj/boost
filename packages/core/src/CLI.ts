/**
 * @copyright   2017-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

// eslint-disable-next-line import/no-extraneous-dependencies
import { Argv } from 'yargs';
import Tool from './Tool';

export default class CLI {
  static registerGlobalOptions(app: Argv, tool: Tool) {
    app
      .option('config', {
        default: '',
        description: tool.msg('app:cliOptionConfig'),
        string: true,
      })
      .option('debug', {
        boolean: true,
        default: false,
        description: tool.msg('app:cliOptionDebug'),
      })
      .option('locale', {
        default: '',
        description: tool.msg('app:cliOptionLocale'),
        string: true,
      })
      .option('output', {
        default: 3,
        description: tool.msg('app:cliOptionOutput'),
        number: true,
      })
      .option('silent', {
        boolean: true,
        default: false,
        description: tool.msg('app:cliOptionSilent'),
      })
      .option('theme', {
        default: 'default',
        description: tool.msg('app:cliOptionTheme'),
        string: true,
      });
  }
}
