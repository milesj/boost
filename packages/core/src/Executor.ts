/**
 * @copyright   2017-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import kebabCase from 'lodash/kebabCase';
import Context from './Context';
import Routine from './Routine';
import Task from './Task';
import Tool from './Tool';
import { Debugger } from './types';

export interface AggregatedResponse {
  errors: Error[];
  results: any[];
}

export default class Executor<Ctx extends Context, Options = {}> {
  context: Ctx;

  debug: Debugger;

  options: Options;

  tool: Tool<any, any>;

  constructor(tool: Tool<any, any>, context: Ctx, options: Partial<Options> = {}) {
    this.context = context;
    this.debug = tool.createDebugger(kebabCase(this.constructor.name));
    this.tool = tool;
    // @ts-ignore
    this.options = { ...options };

    this.debug('Instantiating task executor');
  }

  /**
   * Aggregate and partition errors and results into separate collections.
   */
  aggregateResponse(responses: any[]): AggregatedResponse {
    const results: any[] = [];
    const errors: Error[] = [];

    this.debug('Aggregating results');

    responses.forEach((response: any) => {
      if (response instanceof Error) {
        errors.push(response);
      } else {
        results.push(response);
      }
    });

    return { errors, results };
  }

  /**
   * Execute either a task or routine.
   */
  async execute<T>(
    task: Task<Ctx> | Routine<Ctx, any>,
    value?: T,
    wasParallel: boolean = false,
  ): Promise<any> {
    if (this.getInstanceType(task) === 'Routine') {
      return this.executeRoutine(task as Routine<Ctx, any>, value, wasParallel);
    }

    return this.executeTask(task, value, wasParallel);
  }

  /**
   * Execute a routine with the provided value.
   */
  async executeRoutine<T>(
    routine: Routine<Ctx, any>,
    value?: T,
    wasParallel: boolean = false,
  ): Promise<any> {
    return routine.run(this.context, value, wasParallel);
  }

  /**
   * Execute a task with the provided value.
   */
  async executeTask<T>(task: Task<Ctx>, value?: T, wasParallel: boolean = false): Promise<any> {
    const { console: cli } = this.tool;
    let result = null;

    cli.emit('task', [task, value, wasParallel]);

    try {
      result = await task.run(this.context, value);

      cli.emit('task.pass', [task, result, wasParallel]);
    } catch (error) {
      cli.emit('task.fail', [task, error, wasParallel]);

      throw error;
    }

    return result;
  }

  /**
   * Importing Routine causes a circular reference, so we can't use an instanceof check,
   * so we need to hackily check this another way.
   */
  getInstanceType(task: Task<Ctx> | Routine<Ctx, any>): string {
    let instance = task;
    let name = '';

    do {
      ({ name } = instance.constructor);

      if (name === 'Routine' || name === 'Task') {
        return name;
      }

      instance = Object.getPrototypeOf(instance);
    } while (instance);

    // istanbul ignore next
    return name;
  }

  /**
   * Method to execute tasks. Must be defined in sub-classes.
   */
  async run<T>(tasks: Task<Ctx>[], value?: T): Promise<any> {
    throw new Error('run() must be defined asynchronously.');
  }
}
