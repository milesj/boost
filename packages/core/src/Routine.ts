import execa, { Options as ExecaOptions, ExecaChildProcess, ExecaReturns } from 'execa';
import split from 'split';
import { Readable } from 'stream';
import optimal, { predicates, Blueprint, Predicates } from 'optimal';
import { Event } from '@boost/event';
import Context from './Context';
import Task, { TaskAction } from './Task';
import CoreTool from './Tool';
import { AggregatedResponse } from './Executor';
import ParallelExecutor from './executors/Parallel';
import PoolExecutor, { PoolExecutorOptions } from './executors/Pool';
import SerialExecutor from './executors/Serial';
import SyncExecutor from './executors/Sync';
import instanceOf from './helpers/instanceOf';
import wrapWithPromise from './helpers/wrapWithPromise';
import { STATUS_RUNNING } from './constants';
import { Debugger } from './types';

export interface CommandOptions {
  shell?: boolean;
  task?: Task<any>;
  wrap?: (process: ExecaChildProcess) => void;
}

export default abstract class Routine<
  Ctx extends Context,
  Tool extends CoreTool<any>,
  Options extends object = {}
> extends Task<Ctx> {
  // @ts-ignore Set after instantiation
  debug: Debugger;

  key: string = '';

  onCommand: Event<[string]>;

  onCommandData: Event<[string, string]>;

  options: Options;

  parent: Routine<Ctx, Tool> | null = null;

  // TODO Change to Set in 2.0
  routines: Routine<Ctx, Tool>[] = [];

  // TODO Change to Set in 2.0
  tasks: Task<Ctx>[] = [];

  // @ts-ignore Set after instantiation
  tool: Tool;

  constructor(key: string, title: string, options: Partial<Options> = {}) {
    super(title, (context, value) => this.execute(context, value));

    if (!key || typeof key !== 'string') {
      throw new Error('Routine key must be a valid unique string.');
    }

    this.key = key;
    this.options = optimal(options, this.blueprint(predicates), {
      name: this.constructor.name,
    });

    this.onCommand = new Event('command');
    this.onCommandData = new Event('command.data');
  }

  /**
   * Define an optimal blueprint in which to validate and build the
   * options passed to the constructor.
   */
  blueprint(preds: Predicates): Blueprint<Options> {
    return {} as any;
  }

  /**
   * Called once the routine has been configured and is ready to execute.
   */
  bootstrap() {}

  /**
   * Configure the routine after it has been instantiated.
   */
  configure(parent: Routine<Ctx, Tool>): this {
    this.parent = parent;
    this.tool = parent.tool;
    this.context = parent.context;
    this.metadata.depth = parent.metadata.depth + 1;

    // Custom debugger for this routine
    this.debug = this.tool.createDebugger('routine', this.key);
    this.debug('Bootstrapping routine');

    // Initialize routine
    this.bootstrap();

    return this;
  }

  /**
   * Execute a command with the given arguments and pass the results through a promise.
   */
  async executeCommand(
    command: string,
    args: string[],
    options: ExecaOptions & CommandOptions = {},
  ): Promise<ExecaReturns> {
    const { shell, task, wrap, ...opts } = options;
    const stream = shell
      ? execa.shell(`${command} ${args.join(' ')}`.trim(), opts)
      : execa(command, args, opts);

    this.onCommand.emit([command]);

    // Push chunks to the reporter
    const unit = task || this;
    const handler = (line: string) => {
      if (unit.status === STATUS_RUNNING) {
        unit.output += line;

        // Only capture the status when not empty
        if (line) {
          unit.statusText = line;
        }

        this.onCommandData.emit([command, line]);
      }
    };

    if (!shell) {
      (stream.stdout as Readable).pipe(split()).on('data', handler);
      (stream.stderr as Readable).pipe(split()).on('data', handler);
    }

    // Allow consumer to wrap functionality
    if (typeof wrap === 'function') {
      wrap(stream as ExecaChildProcess);
    }

    return wrapWithPromise(stream);
  }

  /**
   * Execute routines in parallel.
   */
  async parallelizeRoutines<T>(value?: T, routines?: Routine<Ctx, Tool>[]): Promise<any[]> {
    return new ParallelExecutor(this.tool, this.context).runRoutines(
      Array.from(routines || this.routines),
      value,
    );
  }

  /**
   * Execute tasks in parallel.
   */
  async parallelizeTasks<T>(value?: T, tasks?: Task<Ctx>[]): Promise<any[]> {
    return new ParallelExecutor(this.tool, this.context).runTasks(
      Array.from(tasks || this.tasks),
      value,
    );
  }

  /**
   * Add a new routine within this routine.
   */
  pipe(routine: Routine<Ctx, Tool>): this {
    if (instanceOf(routine, Routine)) {
      // eslint-disable-next-line no-param-reassign
      routine.metadata.index = this.routines.length;

      this.routines.push(routine.configure(this));
    } else {
      throw new TypeError(this.tool.msg('errors:routineInstanceInvalid'));
    }

    return this;
  }

  /**
   * Execute routines in a pool.
   */
  async poolRoutines<T>(
    value?: T,
    options?: Partial<PoolExecutorOptions>,
    routines?: Routine<Ctx, Tool>[],
  ): Promise<AggregatedResponse> {
    return new PoolExecutor(this.tool, this.context, options).runRoutines(
      Array.from(routines || this.routines),
      value,
    );
  }

  /**
   * Execute tasks in a pool.
   */
  async poolTasks<T>(
    value?: T,
    options?: Partial<PoolExecutorOptions>,
    tasks?: Task<Ctx>[],
  ): Promise<AggregatedResponse> {
    return new PoolExecutor(this.tool, this.context, options).runTasks(
      Array.from(tasks || this.tasks),
      value,
    );
  }

  /**
   * Execute routines in sequential (serial) order.
   */
  serializeRoutines<T>(value?: T, routines?: Routine<Ctx, Tool>[]): Promise<any> {
    return new SerialExecutor(this.tool, this.context).runRoutines(
      Array.from(routines || this.routines),
      value,
    );
  }

  /**
   * Execute tasks in sequential (serial) order.
   */
  serializeTasks<T>(value?: T, tasks?: Task<Ctx>[]): Promise<any> {
    return new SerialExecutor(this.tool, this.context).runTasks(
      Array.from(tasks || this.tasks),
      value,
    );
  }

  /**
   * Execute routines in sync.
   */
  async synchronizeRoutines<T>(
    value?: T,
    routines?: Routine<Ctx, Tool>[],
  ): Promise<AggregatedResponse> {
    return new SyncExecutor(this.tool, this.context).runRoutines(
      Array.from(routines || this.routines),
      value,
    );
  }

  /**
   * Execute tasks in sync.
   */
  async synchronizeTasks<T>(value?: T, tasks?: Task<Ctx>[]): Promise<AggregatedResponse> {
    return new SyncExecutor(this.tool, this.context).runTasks(
      Array.from(tasks || this.tasks),
      value,
    );
  }

  /**
   * Define an individual task.
   */
  task(title: string, action: TaskAction<Ctx>, scope?: any): Task<Ctx> {
    if (typeof action !== 'function') {
      throw new TypeError(this.tool.msg('errors:taskRequireAction'));
    }

    const task = new Task<Ctx>(title, action.bind(scope || this));

    task.parent = this;
    task.metadata.index = this.tasks.length;

    this.tasks.push(task);

    return task;
  }

  /**
   * Execute the current routine and return a new value.
   * This method *must* be overridden in a subclass.
   */
  abstract async execute(context: Ctx, value: any): Promise<any>;
}
