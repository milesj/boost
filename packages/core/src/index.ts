/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import CLI from './CLI';
import ConfigLoader from './ConfigLoader';
import Console from './Console';
import Context from './Context';
import Emitter, { EventArguments, EventListener } from './Emitter';
import ModuleLoader from './ModuleLoader';
import Pipeline from './Pipeline';
import Plugin from './Plugin';
import Reporter from './Reporter';
import Routine from './Routine';
import Task from './Task';
import Tool, { ToolOptions, ToolConfig, ToolPluginRegistry } from './Tool';

export {
  CLI,
  ConfigLoader,
  Console,
  Context,
  Emitter,
  EventArguments,
  EventListener,
  ModuleLoader,
  Pipeline,
  Plugin,
  Reporter,
  Routine,
  Task,
  Tool,
  ToolConfig,
  ToolOptions,
  ToolPluginRegistry,
};

export * from './types';
