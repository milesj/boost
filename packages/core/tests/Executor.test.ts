import Executor from '../src/Executor';
import Task from '../src/Task';
import Tool from '../src/Tool';
import Routine from '../src/Routine';
import { STATUS_PASSED, STATUS_FAILED } from '../src/constants';
import { createTestTool, createTestRoutine, createTestDebugger } from './helpers';

describe('Executor', () => {
  let tool: Tool;
  let executor: Executor<any>;

  beforeEach(() => {
    tool = createTestTool();

    executor = new Executor(tool, {});
    executor.debug = createTestDebugger();
  });

  describe('aggregateResponse()', () => {
    it('partitions errors and non-errors', () => {
      expect(executor.aggregateResponse([123, new Error(), 'foo', true])).toEqual({
        errors: [new Error()],
        results: [123, 'foo', true],
      });
    });
  });

  describe('execute()', () => {
    it('executes a routine', async () => {
      const spy = jest.fn();
      const routine = createTestRoutine(tool);

      executor.executeRoutine = spy;

      await executor.execute(routine, 123, true);

      expect(spy).toHaveBeenCalledWith(routine, 123, true);
    });

    it('executes a task', async () => {
      const spy = jest.fn();
      const task = new Task('Title', () => Promise.resolve());

      executor.executeTask = spy;

      await executor.execute(task, 123, false);

      expect(spy).toHaveBeenCalledWith(task, 123, false);
    });
  });

  describe('executeRoutine()', () => {
    let routine: Routine<any>;

    beforeEach(() => {
      routine = createTestRoutine(tool);
    });

    it('returns a promise', () => {
      expect(executor.executeRoutine(routine, 123)).toBeInstanceOf(Promise);
    });

    it('passes the value down the promise', async () => {
      expect(await executor.executeRoutine(routine, 123)).toBe(123);
    });

    it('updates status if a success', async () => {
      await executor.executeRoutine(routine, 123);

      expect(routine.status).toBe(STATUS_PASSED);
    });

    it('updates status if a failure', async () => {
      // @ts-ignore
      routine.action = () => {
        throw new Error('Oops');
      };

      try {
        await executor.executeRoutine(routine, 123);
      } catch (error) {
        expect(error).toEqual(new Error('Oops'));
      }

      expect(routine.status).toBe(STATUS_FAILED);
    });
  });

  describe('executeTask()', () => {
    let task: Task<any>;

    beforeEach(() => {
      task = new Task('title', (con, value) => Promise.resolve(value * 3));
    });

    it('returns a promise', () => {
      expect(executor.executeTask(task, 123)).toBeInstanceOf(Promise);
    });

    it('passes the value down the promise', async () => {
      expect(await executor.executeTask(task, 123)).toBe(369);
    });

    it('updates status if a success', async () => {
      await executor.executeTask(task, 123);

      expect(task.status).toBe(STATUS_PASSED);
    });

    it('updates status if a failure', async () => {
      task = new Task('title', () => {
        throw new Error('Oops');
      });

      try {
        await executor.executeTask(task, 123);
      } catch (error) {
        expect(error).toEqual(new Error('Oops'));
      }

      expect(task.status).toBe(STATUS_FAILED);
    });

    it('emits console events if a success', async () => {
      const spy = jest.fn();

      executor.tool.console.emit = spy;

      await executor.executeTask(task, 123);

      expect(spy).toHaveBeenCalledWith('task', [task, 123, false]);
      expect(spy).toHaveBeenCalledWith('task.pass', [task, 369, false]);
    });

    it('emits console events if a failure', async () => {
      const spy = jest.fn();

      executor.tool.console.emit = spy;

      task = new Task('title', () => {
        throw new Error('Oops');
      });

      try {
        await executor.executeTask(task, 123);
      } catch (error) {
        expect(error).toEqual(new Error('Oops'));
      }

      expect(spy).toHaveBeenCalledWith('task', [task, 123, false]);
      expect(spy).toHaveBeenCalledWith('task.fail', [task, new Error('Oops'), false]);
    });

    it('emits console events with parallel flag', async () => {
      const spy = jest.fn();

      executor.tool.console.emit = spy;

      await executor.executeTask(task, 123, true);

      expect(spy).toHaveBeenCalledWith('task', [task, 123, true]);
      expect(spy).toHaveBeenCalledWith('task.pass', [task, 369, true]);
    });
  });

  describe('run()', () => {
    it('errors if not defined', () => {
      expect(executor.run([])).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
