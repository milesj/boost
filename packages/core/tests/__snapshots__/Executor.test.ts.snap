// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Executor run() errors if not defined 1`] = `"run() must be defined asynchronously."`;
